chocolatey/NuGet.Client
# (#16 #36) Add usage of SourceCacheContext everywhere

## Description Of Changes

There are two commits in this PR, one for each of the related problems.

This commit add all required overloads to the methods, as well as adding
entries into the unshipped api files, so that these can be tracked. In
addition, within the LegacyFeedCapabilityResourceV2Feed class a specific
cacheKey was provided for the feed capabilities query, since without
this in place, the HTTP Retry logic path was forced to be executed.

This commit makes changes to support attempting the saving of a cache
file, and it this fails, re-attempt it with a hash of the file path,
which is much shorter, and will hopefully work as expected. To allow
for this to work, it was necessary to change the logic to read the file
again by also checking for the presence of the hashed file name, if the
requested file name doesn't exist.

## Motivation and Context

Two issues are being addressed in this PR:

Within the NuGet.Client assemblies, if there is no SourceCachecContext
available, all outgoing queries are immediately subject to an imposed 1
second delay for each query. When running something like choco outdated
with say 100 packages, this means that this operation will take, as a
minimum, over 100 seconds to complete.

We need to be able to accept a SourceCacheContext for each entry point
so that we can control whether caching happens, or not. For the top
level entry point methods, this is a reasonably simply addition of
adding an overloaded method that takes a SourceCachecContext object.
However, there are other places where adding an overload causes a
cascading of updates due to the inheritance structure of the
ResourceProvider class.  However, it has been decided that these changes
are required, since we need to be able to eliminate all imposed 1 second
delays.

When executing certain queries, it is possible to generate a long file
path, that cannot for saved on certain operating systems. While there
are ways at the OS level to enable this support, we should be able to
handle this without this intervention.

## Testing

1. Check out this [PR](https://github.com/chocolatey/choco/pull/3195) in the chocolatey/choco repository
2. Check out this PR
3. Run the `update-nuget-client` script to bring latest changes from this PR into chocolatey/choco
4. Open fiddler, and have it running throughout the testing
5. Open the chocolaty.log in an application, where you can tail the output
6. Run the following command `choco outdated`
7. Ensure that there are no mention of `Enhanced retry: HttpRetryHandler is in a state that retry would have been abandoned or not waited if it were not enabled` in the log file
8. In fiddler, witness that there are no delays imposed in the outgoing queries
9. Run `choco outdated` again and witness that the command executing again, making only one outgoing query
10. Run `chocolatey search chocolatey` and see that there are outgoing queries, again with no delays
11. Run `choco search chocolatey` again, and see that there is only 1 outgoing query
12. Check the log file, and ensure that there is still no mention of `Enhanced retry: HttpRetryHandler is in a state that retry would have been abandoned or not waited if it were not enabled`

### Operating Systems Testing

Win10

## Change Types Made


## Change Checklist


## Related Issue

- Fixes #16 
- Fixes #36
